staging/src/k8s.io/legacy-cloud-providers/azure/azure_managedDiskController.go

```diff
@@ -85,7 +85,7 @@ func (c *ManagedDiskController) CreateManagedDisk(options *ManagedDiskOptions) (
 	klog.V(4).Infof("azureDisk - creating new managed Name:%s StorageAccountType:%s Size:%v", options.DiskName, options.StorageAccountType, options.SizeGB)
 
 	var createZones *[]string
-	if len(options.AvailabilityZone) > 0 {
+	if len(options.AvailabilityZone) > 0 && options.AvailabilityZone != "0" {
```

Comment:
so `0` means this is not a zone? where is this documented?

Reply:
Strictly speaking this is not documented but I looked over most defined zones in azure and they always start with 1. But when a cluster is created in a region without availability zones or cluster is created on VMs that don't use availability zone feature, kubernetes nodes are being labelled with "0" zone in failure-domain.

Reply:
you mean the value of this `failure-domain.beta.kubernetes.io/zone`? actually it's fault-domain
```
# k get no --show-labels
NAME                           STATUS   ROLES   AGE   VERSION    LABELS
aks-andy-14828428-vmss000004   Ready    agent   67m   v1.15.10   ...,failure-domain.beta.kubernetes.io/region=uksouth,failure-domain.beta.kubernetes.io/zone=0...
aks-andy-14828428-vmss000005   Ready    agent   67m   v1.15.10   ...failure-domain.beta.kubernetes.io/region=uksouth,failure-domain.beta.kubernetes.io/zone=1...
```

related code is here:
https://github.com/feiskyer/kubernetes/blob/7cadf86d75d0eeffefb3e80c5ff6ad151f0412da/staging/src/k8s.io/legacy-cloud-providers/azure/azure_standard.go#L487-L489

and actually `AvailabilityZone` value is passed by `zones` parameter in azure disk storage class, so if user don't want to create a zoned PV, user should not pass value to `zones` parameter

Reply:
For topology aware provisioning I believe zone is automatically picked from selected node even if zones parameter was not present in storage class - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/volume/helpers/zones.go#L121

Is azure different?

Reply:
hi @gnufied per below doc, need to check whether `AvailabilityZone` is a num or not, looks like this is a bug, thanks for finding this issue.

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/docs/using-availability-zones.md#node-labels
Both zoned and unzoned nodes are supported, but the value of node label failure-domain.beta.kubernetes.io/zone are different:

 - For zoned nodes, the value is <region>-<AZ>, e.g. centralus-1.
 - For unzoned nodes, the value is faultDomain, e.g. 0.

Reply:
Are you saying for unzoned nodes - rather than checking for string "0", we should check if `failure-domain.beta.kubernetes.io/zone` is number or not? For unzoned nodes, can this field be anything other than "0" ?


If failure-domain zone can't be anything other than 0 for unzoned nodes then just checking for 0 should be sufficient - isn't it?

Reply:
for unzoned nodes, it's a failure-domain num, it's 0, 1, 2, etc.

Reply:
there is `zoned` parameter to control whether it's a zone cluster or not:
https://github.com/kubernetes/kubernetes/blob/7229e13036e4c89c5b4977ec968229a214ce2d0c/pkg/volume/azure_dd/azure_provision.go#L246-L258

Reply:
Yes but for managed disks it defaults to true even if zone is not specified explicitly. So to clarify - you are proposing that if we detect `failure-domain` label to be a number, we should assume that this is not a zoned cluster and skip using zones? This still leaves the issue with statically provisioned PVs in e2e but I am happy to move the change from checking for 0 to checking for number.

Reply:
if it's 0, then isAvailabilityZone("0") should return "":
https://github.com/kubernetes/kubernetes/blob/7229e13036e4c89c5b4977ec968229a214ce2d0c/staging/src/k8s.io/legacy-cloud-providers/azure/azure_zones.go#L39-L42
why would your test fail?
correct format for selectedAvailabilityZone is `eastus2-1`

Reply:
There are two different failure scenarios. If we specify a empty string zone to `createDisk` call and nodes were in a region without zones then the test fails with:

```
    Retriable: false, RetryAfter: 0s, HTTPStatusCode: 400, RawError: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 400, RawError: {"error":{"code":"LocationNotSupportAvailabilityZones","message":"The resource 'Microsoft.Compute/disks/e2e-7c3fcdda-299f-43f3-a10b-991498f7d5a7' does not support availability zones at location 'westus'."}}
```

Another failure scenario is - if cluster was created in a region with zones but no zone was set then the volume provisioning fails with `InvalidZone '' ` error. Notice the empty string coming from azure.

Reply:
what about fix this like this:
```
diff --git a/staging/src/k8s.io/legacy-cloud-providers/azure/azure_managedDiskController.go b/staging/src/k8s.io/legacy-cloud-providers/azure/azure_managedDiskController.go
index 19049409dd0..ab73965e83b 100644
--- a/staging/src/k8s.io/legacy-cloud-providers/azure/azure_managedDiskController.go
+++ b/staging/src/k8s.io/legacy-cloud-providers/azure/azure_managedDiskController.go
@@ -86,8 +86,11 @@ func (c *ManagedDiskController) CreateManagedDisk(options *ManagedDiskOptions) (

        var createZones *[]string
        if len(options.AvailabilityZone) > 0 {
-               zoneList := []string{c.common.cloud.GetZoneID(options.AvailabilityZone)}
-               createZones = &zoneList
+               zone := c.common.cloud.GetZoneID(options.AvailabilityZone)
+               if zone != "" {
+                       zoneList := []string{zone}
+                       createZones = &zoneList
+               }
        }

        // insert original tags to newTags
```

Reply:
pushed something similar to that. PTAL.